File: backend/src/main/java/com/damiot/backend/model/Device.java
package com.damiot.backend.model;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

/**
 * Entidad JPA para dispositivos ESP32
 * Corresponde a la tabla 'device' en la base de datos
 * 
 * @author Emilio José Salmerón Arjona
 */
@Entity
@Table(name = "device")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Device {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(name = "mac_address", nullable = false, unique = true, length = 17)
    private String macAddress;

    @Column(name = "ip_address", length = 15)
    private String ipAddress;

    @Column(nullable = false, length = 10)
    private String status = "offline"; // "online" u "offline"

    @Column(name = "is_enabled", nullable = false)
    private Boolean isEnabled = true;

    @Column(name = "last_connection")
    private LocalDateTime lastConnection;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (status == null) {
            status = "offline";
        }
        if (isEnabled == null) {
            isEnabled = true;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    /**
     * Verifica si el dispositivo está conectado
     */
    public boolean isOnline() {
        return "online".equalsIgnoreCase(status);
    }

    /**
     * Marca el dispositivo como online
     */
    public void markAsOnline(String ipAddress) {
        this.status = "online";
        this.ipAddress = ipAddress;
        this.lastConnection = LocalDateTime.now();
    }

    /**
     * Marca el dispositivo como offline
     */
    public void markAsOffline() {
        this.status = "offline";
    }
}

File: android/app/src/main/java/com/damiot/android/data/model/Device.kt
package com.damiot.android.data.model

import com.google.gson.annotations.SerializedName

/**
 * Modelo de datos para un dispositivo ESP32
 * 
 * Representa la información de un dispositivo IoT registrado en el sistema.
 * Los campos coinciden con la entidad DeviceStatus del backend.
 * 
 * @property id Identificador único en la base de datos
 * @property name Nombre descriptivo del dispositivo (ej: "ESP32-Salón")
 * @property macAddress Dirección MAC única del ESP32
 * @property ipAddress Dirección IP actual (puede cambiar con DHCP)
 * @property status Estado de conexión: "online" u "offline"
 * @property isEnabled Si el dispositivo está habilitado para mostrarse
 * @property lastConnection Última vez que el dispositivo envió heartbeat
 * 
 * @author Emilio José Salmerón Arjona
 * @since 1.0
 */
data class Device(
    val id: Long,
    val name: String,
    val macAddress: String,
    val ipAddress: String?,
    val status: String,
    @SerializedName("isEnabled")
    val isEnabled: Boolean,
    val lastConnection: String?
) {
    /**
     * Comprueba si el dispositivo está conectado
     * @return true si el estado es "online"
     */
    fun isOnline(): Boolean = status.equals("online", ignoreCase = true)
}
